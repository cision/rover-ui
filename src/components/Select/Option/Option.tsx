import React, { useEffect /*, useMemo */ } from 'react';

// import classNames from 'classnames';

import Dropdown from '../../Dropdown';
// import styles from './Option.module.css';

export interface OptionProps {
  children: React.ReactNode;
  // className?: string;
  disabled?: boolean;
  // onClick?: (arg0?: React.MouseEvent<HTMLButtonElement>) => void;
  // selectedValue?: string;
  value?: string;
}

export const Option: React.FC<OptionProps> = ({
  children = null,
  // className = '',
  // disabled = undefined,
  // selectedValue,
  value: customValue,
  ...props
}) => {
  useEffect(() => {
    if (typeof children !== 'string' && !customValue) {
      console.warn(
        'Option requires a `value` prop when `children` is not a string.'
      );
    }
  }, [children, customValue]);

  // const value = useMemo(
  //   () => customValue || (typeof children === 'string' && children) || '',
  //   [children, customValue]
  // );

  return (
    <Dropdown.Menu.Item
      {
        ...{} /*
        disabled={disabled}
        className={classNames(styles.Option, className, {
          [styles.disabled]: disabled,
          [styles.selected]: selectedValue === value,
        })}
        */
      }
      {...props}
    >
      {children}
    </Dropdown.Menu.Item>
  );
};

export default Option;
