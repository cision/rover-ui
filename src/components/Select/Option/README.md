# \<Option\>

The `Option` component is only to be used as a direct child of `Select`.

The only requirement is that it have a string value for either the `children` or `value` prop, since the string is used as the value of the `Select` when the option is selected.

You may optionally pass a `disabled` prop.
