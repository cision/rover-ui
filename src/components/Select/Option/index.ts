export { default } from './Option';
export type { OptionProps } from './Option';
