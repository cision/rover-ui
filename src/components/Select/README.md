# \<Select\>

Add your documentation for the `Select` component here.

## TODO:

- ~autofocus~
- validity
- ~placeholder - let it unset value!~
- required
- ~on open, scroll to the active or first element (seems to be automatic with focus fixed)~
- keyboard arrow control
- focus
- blur
- max-height
- groups
- typeahead to move the cursor
- set default min width 30 characters and max width ??? - get design input
- _Cut line for MVP_
- set default width to the longest item in the list
- Google autocomplete for native select, maybe punt on it for now
- dropdown open position - auto-detect
  - Detect dropdown menu height (max height included) while invisible
  - Scroll current el to middle of dropdown
  - Put menu offset so selected item is over current input
  - if too close to the bottom edge of window, move window up as far as possible without bumping top edge
- multiselect - needs design work on presenting values and checkboxes or another affordance in the dropdown
- New component: searchable with api callback, and preserve the current selections
