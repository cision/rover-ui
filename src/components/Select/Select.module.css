.Select {
  visibility: visible;
  display: inline-flex;
}

.toggle {
  width: 22ex;
}

.placeholder {
  font-style: italic;
}

.menu {
  max-height: calc(var(--rvr-dropdown-menu-item-height) * var(--rvr-dropdown-menu-max-rows-sm));
}
