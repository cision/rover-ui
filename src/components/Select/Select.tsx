import React, {
  Children,
  forwardRef,
  MutableRefObject,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from 'react';

import type { FC, ReactElement, ReactNode } from 'react';

import classNames from 'classnames';
import { nanoid } from 'nanoid';

import EasyDropdown from '../EasyDropdown';
import type { EasyDropdownProps, MenuItem } from '../EasyDropdown/EasyDropdown';
import styles from './Select.module.css';
import type { OptionProps } from './Option';
import Toggle from './Toggle';

type OptionishElement = ReactElement<OptionProps>;

interface OptionishProps extends OptionProps {
  label: string;
}

const parseOptionProps = ({
  props: { children, value, ...passedProps } = {},
}: ReactElement) =>
  !value && !children
    ? null
    : {
        ...passedProps,
        children,
        label: (typeof children === 'string' && children) || value,
        value,
      };

const makeMenuItemConfigFromChild = ({ handleSelectOption, value }) => (
  item: OptionishProps
) => ({
  ...item,
  onSelect: undefined,
  'aria-selected': value === item.value || value === item.label,
  forwardedRef: React.createRef<HTMLButtonElement>(),
  role: 'option',
  value: item.value === 'string' ? item.value : item.label,
  onClick: () =>
    handleSelectOption(item.value === 'string' ? item.value : item.label),
});

export interface SelectProps
  extends Omit<EasyDropdownProps, 'onChange' | 'placeholder'> {
  autoFocus?: boolean;
  className?: string;
  children?: OptionishElement | OptionishElement[];
  defaultValue?: string;
  disabled?: boolean;
  fauxDisabled?: boolean;
  forwardedRef?: React.Ref<HTMLButtonElement>;
  id?: string;
  onChange?: (arg0: string) => void;
  placeholder?: ReactNode;
  position?: 'top' | 'bottom';
}

const Select: FC<SelectProps> = ({
  autoFocus = false,
  children = undefined,
  className = '',
  defaultValue,
  disabled = false,
  fauxDisabled = false,
  forwardedRef: forwardedToggleRef,
  id: unsafeId = '',
  onChange,
  placeholder = 'Choose something',
  position = 'bottom',
  ...props
}) => {
  const [value, setValue] = useState('');
  const [isOpen, setIsOpen] = useState(false);

  const toggleRef = useRef<HTMLButtonElement>(null) as MutableRefObject<
    HTMLButtonElement
  >;

  useImperativeHandle(forwardedToggleRef, () => toggleRef.current);

  useEffect(() => {
    if (defaultValue !== undefined) setValue(defaultValue);
  }, [defaultValue]);

  const handleSelectOption = useCallback(
    (nextValue) => {
      setValue(nextValue);
      toggleRef?.current?.focus();

      if (onChange) {
        onChange(nextValue);
      }
    },
    [onChange]
  );

  const id = useMemo(() => unsafeId || nanoid(), [unsafeId]);

  const menuItems: MenuItem[] = useMemo(
    () => [
      {
        'aria-selected': value === '',
        className: styles.placeholder,
        forwardedRef: React.createRef<HTMLButtonElement>(),
        children: placeholder,
        label:
          typeof placeholder === 'string' ? placeholder : 'Choose something',
        role: 'option',
        value: '',
        onClick: () => handleSelectOption(''),
      },
      ...Children.map(children || [], parseOptionProps)
        .filter(Boolean)
        .map(makeMenuItemConfigFromChild({ handleSelectOption, value })),
    ],
    [children, handleSelectOption, placeholder, value]
  );

  const handleToggleOpen = useCallback(
    (_: Event, nextIsOpen?: boolean) => {
      setIsOpen(nextIsOpen || false);

      if (!nextIsOpen) {
        toggleRef?.current?.focus();
        return;
      }

      const firstSelected = menuItems.find(
        (i) => value === i.value || value === i.label
      );

      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore: Swear forwardedRef.current is a button here.
      setTimeout(() => firstSelected?.forwardedRef?.current?.focus(), 200);
    },
    [menuItems, value]
  );

  const selectedItem = useMemo(
    () => menuItems.find((i) => value === (i.value || i.label)),
    [menuItems, value]
  );

  const selectionDescriptor =
    selectedItem?.children || selectedItem?.value || placeholder;

  return (
    <EasyDropdown
      {...props}
      className={classNames(styles.Select, className)}
      defaultIsOpen={false}
      menuItems={menuItems}
      onToggle={handleToggleOpen}
      menuProps={{
        role: 'listbox',
        className: styles.menu,
        position: position === 'top' ? 'topRight' : 'bottomRight',
      }}
    >
      <Toggle
        autoFocus={autoFocus}
        id={id}
        isDisabled={disabled || fauxDisabled}
        isOpen={isOpen}
        isPlaceholder={!selectedItem || selectedItem.value === ''}
        label={selectionDescriptor}
        ref={toggleRef}
      />
    </EasyDropdown>
  );
};

export const WithRef = forwardRef<HTMLButtonElement, SelectProps>(
  (props, ref) => <Select {...props} forwardedRef={ref} />
);

export default Select;
