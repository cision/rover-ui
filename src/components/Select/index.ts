export { default } from './Select';
export { default as Option } from './Option';
