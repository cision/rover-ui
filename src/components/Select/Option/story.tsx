import React from 'react';

import { storiesOf } from '@storybook/react';
import { boolean, text } from '@storybook/addon-knobs';

import { Wrap } from '../../../stories/storybook-helpers';

import Option from '.';
import Readme from './README.md';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const FancySpan = (props: Record<string, any>) => <strong {...props} />;

storiesOf('Planets/Select/Option', module)
  .addParameters({
    readme: {
      sidebar: Readme,
    },
  })
  .add(
    'Overview',
    () => {
      const name = text('children', 'Option 1');

      return (
        <Wrap>
          <Option
            disabled={boolean('disabled', false)}
            value={text('value', 'custom value')}
          >
            {name}
          </Option>
        </Wrap>
      );
    },
    {
      info: {
        inline: true,
        source: true,
      },
    }
  )
  .add(
    'Examples',
    () => {
      const now = new Date();

      return (
        <Wrap>
          <Option value={now.toISOString()}>
            Now (time) with an ISO string value
          </Option>
          <Option>Just plain old text</Option>
          <Option value="no-tags-here">
            Don&apos;t put <strong>HTML</strong> or
            <FancySpan>React components</FancySpan> in your option unless you
            specify a value.
          </Option>
          <Option disabled>Disabled</Option>
        </Wrap>
      );
    },
    {
      info: {
        inline: true,
        source: true,
      },
    }
  );
