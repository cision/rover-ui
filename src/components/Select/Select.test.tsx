import React from 'react';
import { render } from '@testing-library/react';

import Select from './Select';
import Option from './Option';

describe('Select', () => {
  it('renders', () => {
    const { baseElement } = render(
      <Select id="foo">
        <Option>Boom</Option>
      </Select>
    );

    expect(baseElement).toMatchSnapshot();
  });
});
