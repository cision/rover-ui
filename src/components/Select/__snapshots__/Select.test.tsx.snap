// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders 1`] = `
<body>
  <div>
    <div
      class="Dropdown Select EasyDropdown"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        class="Input Toggle placeholder"
        data-is-open="false"
        id="foo"
        type="button"
      >
        <span
          class="label"
        >
          Choose something
        </span>
        <svg
          class="icon"
          fill="currentColor"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </button>
    </div>
  </div>
</body>
`;
